File: app/channels/game_channel.rb
class GameChannel < ApplicationCable::Channel
  def subscribed
    @gameHandler = GameServices::GameHandler.new
  end
  def onJoinGame(data)
    game, player, error = @gameHandler.join(data["player_name"], data["room_code"])

    if error.nil?
      stream_for game
      gameJson = game.as_json(include: [:words, :players])
      playerJson = player.as_json
    end

    broadcast_to game, { game: gameJson, player: playerJson, error: error, type: Constants::ActionTypes::GAME_JOINED }
  end

  def onCreateGame(data)
    game, player = @gameHandler.create(data["player_name"], data["word_count"])
    stream_for game

    gameJson = game.as_json(include: [:words, :players])
    playerJson = player.as_json

    broadcast_to game, { game: gameJson, player: playerJson, type: Constants::ActionTypes::GAME_CREATED }
  end

  def onGameStart()
    @gameHandler.start()
  end

  def onAnswer(data)
    @gameHandler.handleAnswer(
      data["player_id"],
      data["answer"]
    )
  end

  def onVote(data)
    @gameHandler.handleVote(
      data["player_id"],
      data["definition"],
      data["voted_for_id"]
    )
  end

  def onNextWord()
    @gameHandler.nextWord()
  end

  def onGameRoomClose(data)
    @gameHandler.closeGameRoom()
  end
end

File: app/javascript/channels/game_channel.ts
import consumer from "./consumer"
import { CREATE_GAME } from "../store/actions/actionTypes";
import { gameCreated } from "../store/actions/game";
import store from "../store/index";

consumer.subscriptions.create("GameChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
    console.log("Connection succeeded.");
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
    console.log("Disconnected from server.");
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    store.dispatch({type: CREATE_GAME, payload: data});
  },
  
  joinGame(player_name, room_code) {
    this.perform("joinGame", {"player_name": player_name, "room_code": room_code});
  },

  createGame(player_name, word_count) {
    this.perform("createGame", {"player_name": player_name, "word_count": word_count});
  },

  onAnswer(data)  {
    this.perform("onAnswer", data);
  },

  onVote(data) {
    this.perform("onVote", data);
  },

  onGameEnd(data) {
    this.perform("onGameEnd", data);
  }

});
